src/protobuf-net.Reflection/CSharpCodeGenerator.cs

```diff
@@ -839,23 +868,76 @@ protected string MakeRelativeName(GeneratorContext ctx, string typeName)
+            ref NullabilityType? nullabilityType,
+            ref CompatibilityLevel? compatibilityLevel,
+            ref bool isMap,
             bool nonNullable = false)
         {
+            if (ctx.EmitNullWrappers)
+            {
+                switch (typeName)
+                {
+                    case WellKnownTypeDouble:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "double" : "double?";
+                    case WellKnownTypeFloat:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "float" : "float?";
+                    case WellKnownTypeInt64:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "long" : "long?";
+                    case WellKnownTypeUInt64:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "ulong" : "ulong?";
+                    case WellKnownTypeInt32:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "int" : "int?";
+                    case WellKnownTypeUInt32:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "uint" : "uint?";
+                    case WellKnownTypeBool:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "bool" : "bool?";
+                    case WellKnownTypeString:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return "string";
+                    case WellKnownTypeBytes:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return "byte[]";
+                }
+            }
+            
             switch (typeName)
             {
+                #region WellKnownTypes
+
                 case WellKnownTypeTimestamp:
-                    dataFormat = "WellKnown";
+                    if (ctx.EmitCompatibilityLevelAttribute) compatibilityLevel = CompatibilityLevel.Level300;
+                    else dataFormat = "WellKnown";
                     return nonNullable ? "global::System.DateTime" : "global::System.DateTime?";
                 case WellKnownTypeDuration:
-                    dataFormat = "WellKnown";
+                    if (ctx.EmitCompatibilityLevelAttribute) compatibilityLevel = CompatibilityLevel.Level300;
+                    else dataFormat = "WellKnown";
                     return nonNullable ? "global::System.TimeSpan" : "global::System.TimeSpan?";
+                
                 case WellKnownTypeEmpty:
                     return "global::ProtoBuf.Empty";
+                
+                #endregion
+
+                #region BCL
```

Comment:
I'm not sure we're adding much with a `#region` here; can we just `// BCL` and `// well-known types` ?

Reply:
leftover after previous implementation approach, removed